Modifies USBDaemon to act as a server

johnnyman727 opened the pull request:
Fixes https://github.com/tessel/t2-firmware/issues/132.

So far I investigated how to create a domain socket that will act as a server using `epoll` and how to create a domain socket that will act as a client using `poll`. The reason we're using two different types of non-blocking polling is because the usb daemon could have an unknown number of concurrent connections and `epoll` is more efficient because the mechanics allow you to check only those events that fired. `poll`, on the other hand, is better for the SPI Daemon because it will always have four possible file descriptors (IRQ, USB Daemon, MOD PORT A, MOD PORT B) and the interface is somewhat simpler.

The snippets below basically have the usb daemon waiting for a new connection, accepting a new socket (the spi daemon), reading 4 bytes from the socket, writing 4 bytes to the socket, then handling the closing of the socket and going back into accepting new connections.

The next step is to integrate these simple cases into the existing usb and spi daemons.

LinusU commented on soc/spid-slim.c:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    return 0;
+}
```
It's best practice to always include a newline at the end of each file

johnnyman727 replied:
Thanks! Fixed in my next commit.

johnnyman727 commented:
@kevinmehall If you get a chance, could you review these changes?

I modified the USB Daemon to act as a server to a single incoming connection and wipe any existing processes upon disconnect. I also refactored it a bit but in any case, it's straightforward and I feel good about it.

The SPI Daemon now uses the `CONN_POLL` of the USB channel as a client that connects to the USB Daemon when the channel is enabled by the microcontroller (the `SOCK_POLL` entry is left empty). This modification is a bit more of a hack job so let me know if you have refactoring suggestions. Also, the USB channel socket gets recreated each time the channel is enabled, in contrast to the module port sockets which stay constant the entire script. When I tried to create a single socket instance that gets closed and opened, I had issues with the `POLLHUP` flag being set immediately (before the channel was enabled) which then closed the connection (because the socket wasn't connected yet). That was an issue because when the channel *did* get enabled, the `fd` was already -1 because it was closed previously. Happy to hear your comments on that as well.

It seems to be working well from my basic tests. The CLI still has issues with [like this one](https://github.com/tessel/t2-cli/issues/477) where the USB interface isn't released which is mucking some things up and it definitely doesn't release the interface upon a crash. I'll be looking into that next.
